File: services/event-streaming-service/src/storage/iceberg.rs
//! Iceberg-backed dataset writer.
//!
//! The crate currently does **not** depend on a fully-featured Iceberg
//! implementation: the real REST Catalog client and Parquet/manifest writers
//! are scoped to *Tarea 1.1* in `libs/storage-abstraction` and will be wired
//! in later. To keep this crate functional and testable in the meantime, the
//! writer is split in two collaborators:
//!
//! * [`IcebergCatalog`] — a small trait that captures the only interaction we
//!   need today: registering a new snapshot for an `(namespace, table)`
//!   pair. [`InMemoryCatalog`] is the test double; [`RestCatalogClient`] is a
//!   thin HTTP shim against an Iceberg REST Catalog endpoint
//!   (`POST /v1/namespaces/{ns}/tables/{table}/snapshots`).
//! * The writer itself, which uploads the snapshot bytes to the configured
//!   [`storage_abstraction::StorageBackend`] under an Iceberg-shaped path
//!   (`<namespace>/<table>/data/<snapshot_id>.parquet`) and then asks the
//!   catalog to commit the new snapshot pointing at that data file.
//!
//! When the real Iceberg backend lands in `storage-abstraction` (gated by the
//! `iceberg` feature), this module will switch to that implementation while
//! keeping the same public surface and tests.

use std::collections::HashMap;
use std::sync::{Arc, Mutex};

use storage_abstraction::StorageBackend;

use super::writer::{DatasetSnapshot, DatasetWriter, WriteOutcome, WriterError};

/// Fully-qualified Iceberg table reference (namespace + table name).
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct IcebergTableRef {
    pub namespace: String,
    pub table: String,
}

impl IcebergTableRef {
    pub fn new(namespace: impl Into<String>, table: impl Into<String>) -> Self {
        Self {
            namespace: namespace.into(),
            table: table.into(),
        }
    }
}

/// A single committed snapshot as known to the catalog.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SnapshotCommit {
    pub snapshot_id: String,
    pub data_file: String,
    pub summary: serde_json::Value,
}

/// Minimal abstraction over an Iceberg catalog. The trait deliberately covers
/// only the operations needed by the writer so it can be backed by either an
/// in-memory mock (tests) or the REST Catalog (production).
#[async_trait::async_trait]
pub trait IcebergCatalog: Send + Sync + std::fmt::Debug {
    /// Append a new snapshot to the table. Implementations are expected to
    /// create the table on first use.
    async fn append_snapshot(
        &self,
        table: &IcebergTableRef,
        commit: SnapshotCommit,
    ) -> Result<(), WriterError>;
}

/// In-memory catalog used by unit tests and as a fallback when no REST
/// Catalog endpoint is configured but the operator still wants to exercise
/// the Iceberg code path locally.
#[derive(Debug, Default, Clone)]
pub struct InMemoryCatalog {
    inner: Arc<Mutex<HashMap<IcebergTableRef, Vec<SnapshotCommit>>>>,
}

impl InMemoryCatalog {
    pub fn new() -> Self {
        Self::default()
    }

    /// Returns a clone of the snapshots committed for a given table.
    pub fn snapshots(&self, table: &IcebergTableRef) -> Vec<SnapshotCommit> {
        self.inner
            .lock()
            .expect("InMemoryCatalog mutex poisoned")
            .get(table)
            .cloned()
            .unwrap_or_default()
    }
}

#[async_trait::async_trait]
impl IcebergCatalog for InMemoryCatalog {
    async fn append_snapshot(
        &self,
        table: &IcebergTableRef,
        commit: SnapshotCommit,
    ) -> Result<(), WriterError> {
        let mut guard = self
            .inner
            .lock()
            .map_err(|_| WriterError::Catalog("in-memory catalog mutex poisoned".to_string()))?;
        guard.entry(table.clone()).or_default().push(commit);
        Ok(())
    }
}

/// HTTP client for an Iceberg REST Catalog endpoint.
///
/// This is intentionally a thin shim: it sends the snapshot summary as JSON
/// and trusts the server to perform the actual table mutation. When the full
/// Iceberg client lands in `libs/storage-abstraction` (Tarea 1.1) this type
/// will delegate to it instead of issuing the HTTP call directly.
#[derive(Debug, Clone)]
pub struct RestCatalogClient {
    base_url: String,
    http: reqwest::Client,
}

impl RestCatalogClient {
    pub fn new(base_url: impl Into<String>) -> Self {
        Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            http: reqwest::Client::new(),
        }
    }

    pub fn with_client(base_url: impl Into<String>, http: reqwest::Client) -> Self {
        Self {
            base_url: base_url.into().trim_end_matches('/').to_string(),
            http,
        }
    }

    pub fn base_url(&self) -> &str {
        &self.base_url
    }
}

#[async_trait::async_trait]
impl IcebergCatalog for RestCatalogClient {
    async fn append_snapshot(
        &self,
        table: &IcebergTableRef,
        commit: SnapshotCommit,
    ) -> Result<(), WriterError> {
        let url = format!(
            "{}/v1/namespaces/{}/tables/{}/snapshots",
            self.base_url, table.namespace, table.table
        );
        let body = serde_json::json!({
            "snapshot-id": commit.snapshot_id,
            "data-file": commit.data_file,
            "summary": commit.summary,
            "operation": "append",
        });
        let response = self
            .http
            .post(&url)
            .json(&body)
            .send()
            .await
            .map_err(|e| WriterError::Catalog(format!("REST catalog request failed: {e}")))?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            return Err(WriterError::Catalog(format!(
                "REST catalog returned {status}: {text}"
            )));
        }
        Ok(())
    }
}

/// Iceberg-backed [`DatasetWriter`].
#[derive(Clone)]
pub struct IcebergDatasetWriter {
    backend: Arc<dyn StorageBackend>,
    catalog: Arc<dyn IcebergCatalog>,
    namespace: String,
}

impl IcebergDatasetWriter {
    pub fn new(
        backend: Arc<dyn StorageBackend>,
        catalog: Arc<dyn IcebergCatalog>,
        namespace: impl Into<String>,
    ) -> Self {
        Self {
            backend,
            catalog,
            namespace: namespace.into(),
        }
    }

    fn data_path(&self, snapshot: &DatasetSnapshot) -> String {
        format!(
            "{}/{}/data/{}.parquet",
            self.namespace, snapshot.table, snapshot.snapshot_id
        )
    }

    fn iceberg_uri(&self, snapshot: &DatasetSnapshot) -> String {
        format!(
            "iceberg://{}/{}#{}",
            self.namespace, snapshot.table, snapshot.snapshot_id
        )
    }
}

impl std::fmt::Debug for IcebergDatasetWriter {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("IcebergDatasetWriter")
            .field("namespace", &self.namespace)
            .field("catalog", &self.catalog)
            .finish()
    }
}

#[async_trait::async_trait]
impl DatasetWriter for IcebergDatasetWriter {
    fn backend_name(&self) -> &'static str {
        "iceberg"
    }

    async fn append(&self, snapshot: DatasetSnapshot) -> Result<WriteOutcome, WriterError> {
        if snapshot.table.is_empty() || snapshot.snapshot_id.is_empty() {
            return Err(WriterError::InvalidSnapshot(
                "table and snapshot_id are required".to_string(),
            ));
        }

        let data_path = self.data_path(&snapshot);
        self.backend
            .put(&data_path, snapshot.payload.clone())
            .await?;

        let table_ref = IcebergTableRef::new(self.namespace.clone(), snapshot.table.clone());
        let commit = SnapshotCommit {
            snapshot_id: snapshot.snapshot_id.clone(),
            data_file: data_path,
            summary: snapshot.metadata.clone(),
        };
        self.catalog.append_snapshot(&table_ref, commit).await?;

        Ok(WriteOutcome {
            backend: "iceberg",
            location: self.iceberg_uri(&snapshot),
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use bytes::Bytes;
    use storage_abstraction::local::LocalStorage;

    fn fixture() -> (tempfile::TempDir, Arc<dyn StorageBackend>, Arc<InMemoryCatalog>) {
        let dir = tempfile::tempdir().unwrap();
        let store = LocalStorage::new(dir.path().to_str().unwrap()).unwrap();
        (dir, Arc::new(store), Arc::new(InMemoryCatalog::new()))
    }

    #[tokio::test]
    async fn append_writes_data_file_and_commits_snapshot() {
        let (_guard, backend, catalog) = fixture();
        let writer = IcebergDatasetWriter::new(
            backend.clone(),
            catalog.clone(),
            "streaming_service",
        );

        let snapshot = DatasetSnapshot::new("window_42", "snap_001", Bytes::from_static(b"row"))
            .with_metadata(serde_json::json!({"rows": 1}));
        let outcome = writer.append(snapshot).await.unwrap();

        assert_eq!(outcome.backend, "iceberg");
        assert_eq!(outcome.location, "iceberg://streaming_service/window_42#snap_001");
        assert!(
            backend
                .exists("streaming_service/window_42/data/snap_001.parquet")
                .await
                .unwrap()
        );
        let commits = catalog.snapshots(&IcebergTableRef::new("streaming_service", "window_42"));
        assert_eq!(commits.len(), 1);
        assert_eq!(commits[0].snapshot_id, "snap_001");
        assert_eq!(commits[0].data_file, "streaming_service/window_42/data/snap_001.parquet");
    }

    #[tokio::test]
    async fn append_is_idempotent_per_snapshot_id_at_storage_layer() {
        // The catalog records every commit; the storage layer overwrites the
        // data file. Both calls must succeed: replay safety lives at the
        // catalog level (not asserted here), but we want no spurious errors.
        let (_guard, backend, catalog) = fixture();
        let writer = IcebergDatasetWriter::new(backend, catalog.clone(), "streaming_service");

        let snap = DatasetSnapshot::new("window_x", "snap_1", Bytes::from_static(b"v1"));
        writer.append(snap.clone()).await.unwrap();
        writer.append(snap).await.unwrap();

        let commits = catalog.snapshots(&IcebergTableRef::new("streaming_service", "window_x"));
        assert_eq!(commits.len(), 2);
    }

    #[tokio::test]
    async fn append_rejects_empty_identifiers() {
        let (_guard, backend, catalog) = fixture();
        let writer = IcebergDatasetWriter::new(backend, catalog, "streaming_service");
        let snap = DatasetSnapshot::new("window_x", "", Bytes::new());
        assert!(matches!(
            writer.append(snap).await.unwrap_err(),
            WriterError::InvalidSnapshot(_)
        ));
    }

    #[test]
    fn rest_catalog_client_trims_trailing_slash() {
        let client = RestCatalogClient::new("http://catalog.local:8181/");
        assert_eq!(client.base_url(), "http://catalog.local:8181");
    }
}

File: openfoundry-go/services/ingestion-replication-service/internal/storage/iceberg.go
package storage

// Iceberg-backed dataset writer.
//
// The service currently does not depend on a fully-featured Iceberg
// implementation: the real REST Catalog client and Parquet/manifest writers
// are scoped to the libs/storage-abstraction port and will be wired in later.
// To keep this package functional and testable in the meantime, the writer is
// split in two collaborators:
//
//   - IcebergCatalog: a small interface that captures the only interaction we
//     need today, registering a new snapshot for an (namespace, table) pair.
//     InMemoryCatalog is the test double; RestCatalogClient is a thin HTTP
//     shim against an Iceberg REST Catalog endpoint
//     (POST /v1/namespaces/{ns}/tables/{table}/snapshots).
//   - The writer itself, which uploads the snapshot bytes to the configured
//     StorageBackend under an Iceberg-shaped path
//     (<namespace>/<table>/data/<snapshot_id>.parquet) and then asks the
//     catalog to commit the new snapshot pointing at that data file.

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
)

// IcebergTableRef is a fully-qualified Iceberg table reference (namespace +
// table name).
type IcebergTableRef struct {
	Namespace string
	Table     string
}

// NewIcebergTableRef constructs an IcebergTableRef. Mirrors
// IcebergTableRef::new in Rust.
func NewIcebergTableRef(namespace, table string) IcebergTableRef {
	return IcebergTableRef{Namespace: namespace, Table: table}
}

// SnapshotCommit is a single committed snapshot as known to the catalog.
type SnapshotCommit struct {
	SnapshotID string
	DataFile   string
	Summary    json.RawMessage
}

// Equal reports whether two commits are byte-equal in their identifying
// fields and in their serialised summary.
func (c SnapshotCommit) Equal(other SnapshotCommit) bool {
	if c.SnapshotID != other.SnapshotID || c.DataFile != other.DataFile {
		return false
	}
	return bytes.Equal(c.Summary, other.Summary)
}

// IcebergCatalog is the minimal abstraction over an Iceberg catalog. The
// interface deliberately covers only the operations needed by the writer so
// it can be backed by either an in-memory mock (tests) or the REST Catalog
// (production).
type IcebergCatalog interface {
	// AppendSnapshot appends a new snapshot to the table. Implementations
	// are expected to create the table on first use.
	AppendSnapshot(ctx context.Context, table IcebergTableRef, commit SnapshotCommit) error
}

// InMemoryCatalog is the in-memory IcebergCatalog used by unit tests and as a
// fallback when no REST Catalog endpoint is configured but the operator still
// wants to exercise the Iceberg code path locally.
type InMemoryCatalog struct {
	mu    sync.Mutex
	inner map[IcebergTableRef][]SnapshotCommit
}

// NewInMemoryCatalog returns an empty in-memory catalog.
func NewInMemoryCatalog() *InMemoryCatalog {
	return &InMemoryCatalog{inner: make(map[IcebergTableRef][]SnapshotCommit)}
}

// AppendSnapshot implements IcebergCatalog.
func (c *InMemoryCatalog) AppendSnapshot(_ context.Context, table IcebergTableRef, commit SnapshotCommit) error {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.inner[table] = append(c.inner[table], commit)
	return nil
}

// Snapshots returns a copy of the snapshots committed for the given table.
func (c *InMemoryCatalog) Snapshots(table IcebergTableRef) []SnapshotCommit {
	c.mu.Lock()
	defer c.mu.Unlock()
	src, ok := c.inner[table]
	if !ok {
		return nil
	}
	out := make([]SnapshotCommit, len(src))
	copy(out, src)
	return out
}

// HTTPDoer is the slice of *http.Client RestCatalogClient depends on. Lets
// tests inject a fake transport without spinning up a real server.
type HTTPDoer interface {
	Do(req *http.Request) (*http.Response, error)
}

// RestCatalogClient is the HTTP client for an Iceberg REST Catalog endpoint.
//
// This is intentionally a thin shim: it sends the snapshot summary as JSON
// and trusts the server to perform the actual table mutation. When the full
// Iceberg client lands in libs/storage-abstraction, this type will delegate
// to it instead of issuing the HTTP call directly.
type RestCatalogClient struct {
	baseURL string
	http    HTTPDoer
}

// NewRestCatalogClient builds a client targeting baseURL. Trailing slashes are
// trimmed, matching RestCatalogClient::new in Rust.
func NewRestCatalogClient(baseURL string) *RestCatalogClient {
	return &RestCatalogClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		http:    &http.Client{},
	}
}

// NewRestCatalogClientWithHTTP builds a client with a caller-supplied HTTP
// doer. Mirrors RestCatalogClient::with_client in Rust and is the test seam
// for swapping in a stub transport.
func NewRestCatalogClientWithHTTP(baseURL string, doer HTTPDoer) *RestCatalogClient {
	return &RestCatalogClient{
		baseURL: strings.TrimRight(baseURL, "/"),
		http:    doer,
	}
}

// BaseURL returns the trimmed base URL the client posts to.
func (c *RestCatalogClient) BaseURL() string { return c.baseURL }

// restCatalogBody mirrors the JSON shape Rust emits via serde_json::json!.
// Field declaration order is preserved by encoding/json so the body is
// byte-equivalent to the Rust client.
type restCatalogBody struct {
	SnapshotID string          `json:"snapshot-id"`
	DataFile   string          `json:"data-file"`
	Summary    json.RawMessage `json:"summary"`
	Operation  string          `json:"operation"`
}

// AppendSnapshot implements IcebergCatalog by POSTing the commit to the REST
// catalog endpoint:
//
//	POST <base>/v1/namespaces/{ns}/tables/{table}/snapshots
func (c *RestCatalogClient) AppendSnapshot(ctx context.Context, table IcebergTableRef, commit SnapshotCommit) error {
	url := fmt.Sprintf("%s/v1/namespaces/%s/tables/%s/snapshots", c.baseURL, table.Namespace, table.Table)
	summary := commit.Summary
	if len(summary) == 0 {
		summary = json.RawMessage("null")
	}
	body, err := json.Marshal(restCatalogBody{
		SnapshotID: commit.SnapshotID,
		DataFile:   commit.DataFile,
		Summary:    summary,
		Operation:  "append",
	})
	if err != nil {
		return NewCatalogError("REST catalog request failed: %s", err.Error())
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return NewCatalogError("REST catalog request failed: %s", err.Error())
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.http.Do(req)
	if err != nil {
		return NewCatalogError("REST catalog request failed: %s", err.Error())
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		text, _ := io.ReadAll(resp.Body)
		return NewCatalogError("REST catalog returned %d %s: %s", resp.StatusCode, http.StatusText(resp.StatusCode), string(text))
	}
	return nil
}

// IcebergDatasetWriter is the Iceberg-backed DatasetWriter.
type IcebergDatasetWriter struct {
	backend   StorageBackend
	catalog   IcebergCatalog
	namespace string
}

// NewIcebergDatasetWriter wires the writer with the given backend, catalog
// and namespace. Mirrors IcebergDatasetWriter::new in Rust.
func NewIcebergDatasetWriter(backend StorageBackend, catalog IcebergCatalog, namespace string) *IcebergDatasetWriter {
	return &IcebergDatasetWriter{backend: backend, catalog: catalog, namespace: namespace}
}

// Namespace returns the configured catalog namespace.
func (w *IcebergDatasetWriter) Namespace() string { return w.namespace }

func (w *IcebergDatasetWriter) dataPath(snapshot DatasetSnapshot) string {
	return fmt.Sprintf("%s/%s/data/%s.parquet", w.namespace, snapshot.Table, snapshot.SnapshotID)
}

func (w *IcebergDatasetWriter) icebergURI(snapshot DatasetSnapshot) string {
	return fmt.Sprintf("iceberg://%s/%s#%s", w.namespace, snapshot.Table, snapshot.SnapshotID)
}

// BackendName implements DatasetWriter.
func (w *IcebergDatasetWriter) BackendName() string { return "iceberg" }

// Append implements DatasetWriter. The bytes are uploaded to the configured
// StorageBackend at <namespace>/<table>/data/<snapshot_id>.parquet, then the
// catalog is asked to commit the new snapshot pointing at that data file.
func (w *IcebergDatasetWriter) Append(ctx context.Context, snapshot DatasetSnapshot) (WriteOutcome, error) {
	if snapshot.Table == "" || snapshot.SnapshotID == "" {
		return WriteOutcome{}, NewInvalidSnapshotError("table and snapshot_id are required")
	}

	dataPath := w.dataPath(snapshot)
	if err := w.backend.Put(ctx, dataPath, snapshot.Payload); err != nil {
		return WriteOutcome{}, NewStorageError(err)
	}

	tableRef := NewIcebergTableRef(w.namespace, snapshot.Table)
	commit := SnapshotCommit{
		SnapshotID: snapshot.SnapshotID,
		DataFile:   dataPath,
		Summary:    snapshot.Metadata,
	}
	if err := w.catalog.AppendSnapshot(ctx, tableRef, commit); err != nil {
		return WriteOutcome{}, err
	}

	return WriteOutcome{Backend: "iceberg", Location: w.icebergURI(snapshot)}, nil
}
